services/galley/src/Galley/API/Update.hs

```diff
@@ -885,60 +876,32 @@ rmBot zusr zcon b = do
 -------------------------------------------------------------------------------
 -- Helpers
 
-addToConversation :: ([BotMember], [LocalMember]) -> [RemoteMember] -> (UserId, RoleName) -> ConnId -> [(UserId, RoleName)] -> [(Remote UserId, RoleName)] -> Data.Conversation -> Galley UpdateResult
+addToConversation ::
+  -- | The existing bots and local users in the conversation
+  ([BotMember], [LocalMember]) ->
+  -- | The existing remote users
+  [RemoteMember] ->
+  -- | The originating user and their role
+  (UserId, RoleName) ->
+  -- | The connection ID of the originating user
+  ConnId ->
+  -- | New local users to be added and their roles
+  [(UserId, RoleName)] ->
+  -- | New remote users to be added and their roles
+  [(Remote UserId, RoleName)] ->
+  -- | The conversation to modify
+  Data.Conversation ->
+  Galley UpdateResult
 addToConversation _ _ _ _ [] [] _ = pure Unchanged
-addToConversation (bots, lothers) rothers (usr, usrRole) conn locals remotes c = do
+addToConversation (bots, existingLocals) existingRemotes (usr, usrRole) conn newLocals newRemotes c = do
   ensureGroupConv c
-  mems <- checkedMemberAddSize locals remotes
+  mems <- checkedMemberAddSize newLocals newRemotes
   now <- liftIO getCurrentTime
   localDomain <- viewFederationDomain
   (e, lmm, rmm) <- Data.addMembersWithRole localDomain now (Data.convId c) (usr, usrRole) mems
-  let mm = catMembers localDomain lmm rmm
-      qcnv = Qualified (Data.convId c) localDomain
-      qusr = Qualified usr localDomain
-  -- FUTUREWORK: parallelise federated requests
-  traverse_ (uncurry (updateRemoteConversations now mm qusr qcnv))
-    . Map.assocs
-    . partitionQualified
-    . nubOrd
-    . map (unTagged . rmId)
-    $ rmm <> rothers
-  let allMembers = nubOrdOn memId (lmm <> lothers)
-  for_ (newPush ListComplete usr (ConvEvent e) (recipient <$> allMembers)) $ \p ->
-    push1 $ p & pushConn ?~ conn
-  void . forkIO $ void $ External.deliver (bots `zip` repeat e)
+  notifyRemotes existingRemotes usr now c lmm rmm
```

Comment:
Note that `notifyRemotes` not only results in remote users getting notifications, but also changes state on the remote end in the form of database updates.

I think the naming of these two functions should reflect the fact that they have sightly different side effects, and their difference is not only in terms of operating locally vs remotely.